compiler/damlc/lib/DA/Cli/Damlc/Test.hs

```diff
@@ -91,36 +101,49 @@ failedTestOutput h file = do
     pure $ map (, Just errMsg) $ fromMaybe [VRScenario file "Unknown"] mbScenarioNames
 
 
-printTestCoverage :: [LF.Module] -> [(VirtualResource, Either SSC.Error SS.ScenarioResult)] -> IO ()
-printTestCoverage dalfs results
+printTestCoverage ::
+       ShowCoverage
+    -> [LF.Module]
+    -> [(VirtualResource, Either SSC.Error SS.ScenarioResult)]
+    -> IO ()
+printTestCoverage ShowCoverage {getShowCoverage} dalfs results
   | any (\(_, errOrRes) -> isLeft errOrRes) results = pure ()
-  | otherwise =
-    putStrLn $
-    unwords
-        [ "test coverage: templates"
-        , percentage coveredNrOfTemplates nrOfTemplates <> ","
-        , "choices"
-        , percentage coveredNrOfChoices nrOfChoices
-        ]
+  | otherwise = do
+      putStrLn $
+          unwords
+              [ "test coverage: templates"
+              , percentage coveredNrOfTemplates nrOfTemplates <> ","
+              , "choices"
+              , percentage coveredNrOfChoices nrOfChoices
+              ]
+      when getShowCoverage $ do
+          putStrLn $
+              unlines $
+              ["templates never created:"] <> map T.unpack missingTemplates <>
+              ["choices never executed:"] <>
+              map T.unpack missingChoices
   where
-    templates = [t | m <- dalfs , t <- NM.toList $ LF.moduleTemplates m]
-    nrOfTemplates = length templates
-    nrOfChoices = length [n | t <- templates, n <- NM.names $ LF.tplChoices t]
+    templates = [(m, t) | m <- dalfs , t <- NM.toList $ LF.moduleTemplates m]
+    choices = [(m, t, n) | (m, t) <- templates, n <- NM.names $ LF.tplChoices t]
     percentage i j
       | j > 0 = show (round @Double $ 100.0 * (fromIntegral i / fromIntegral j) :: Int) <> "%"
       | otherwise = "100%"
     allScenarioNodes = [n | (_vr, Right res) <- results, n <- V.toList $ SS.scenarioResultNodes res]
-    coveredNrOfTemplates =
-        length $
+    coveredTemplates =
         nubSort $
         [ SS.contractInstanceTemplateId contractInstance
         | n <- allScenarioNodes
         , Just (SS.NodeNodeCreate SS.Node_Create {SS.node_CreateContractInstance}) <-
               [SS.nodeNode n]
         , Just contractInstance <- [node_CreateContractInstance]
         ]
-    coveredNrOfChoices =
-        length $
+    missingTemplates =
+        [ (LF.moduleNameString $ LF.moduleName m) <> ":" <>
+        (T.concat $ LF.unTypeConName $ LF.tplTypeCon t)
+        | (m, t) <- templates
+        ] \\
```

Comment:
Let’s turn this into a set lookup. I don’t want to end up on accidentally quadratic :)

Reply:
turned it to sets.